x/bridge/keeper/transfers.go

```diff
@@ -31,90 +26,96 @@ func (k Keeper) InboundTransfer(
+	if !found {
 		return errorsmod.Wrapf(types.ErrInvalidAssetID, "Asset not found %s", assetID.Name())
 	}
 	if !asset.Status.InboundActive() {
 		return errorsmod.Wrapf(types.ErrInvalidAssetStatus, "Inbound transfers are disabled for this asset")
 	}
 
-	// Try to finalize the transfer in store
-	err := k.finalizeInboundTransfer(ctx, externalID, sender, destAddr, assetID, amount, params.VotesNeeded)
-	switch {
-	case err == nil:
-		// Everything is fine, mint!
-	case errors.Is(err, ErrAlreadyFinalized), errors.Is(err, ErrNeedMoreVotes):
-		// Expected scenario, just return without minting
-		return nil
-	default:
-		// Unexpected error
+	// Try to finalize the transfer
+	finalized, err := k.finalizeInboundTransfer(
+		ctx,
+		externalID,
+		externalHeight,
+		sender,
+		destAddr,
+		assetID,
+		amount,
+		params.VotesNeeded,
+	)
+	if err != nil {
 		return errorsmod.Wrapf(sdkerrors.ErrLogic, "Can't finalize inbound trander: %s", err.Error())
 	}
+	if !finalized {
+		// The transfer either doesn't have enough votes or has already been finalised
+		return nil
+	}
 
+	// Perform tokenfactory mint
 	err = k.mint(ctx, destAddr, assetID, amount)
 	if err != nil {
 		return errorsmod.Wrap(types.ErrTokenfactory, err.Error())
 	}
 
+	// Now when the transfer is finalized, update the state
+	k.SaveFinalizedTransfer(ctx, externalID)
+	k.UpdateLastAssetHeight(ctx, assetID, externalHeight)
+
 	return nil
 }
 
+// finalizeInboundTransfer returns true if the transfer is successfully finalized,
+// i.e., the transfer was not finalized before adding the new voter to the voter list,
+// but is finalized after the addition.
 func (k Keeper) finalizeInboundTransfer(
 	ctx sdk.Context,
 	externalID string,
+	externalHeight uint64,
 	sender string,
 	destAddr string,
 	assetID types.AssetID,
 	amount math.Int,
 	votesNeeded uint64,
-) error {
+) (bool, error) {
+	if k.IsTransferFinalized(ctx, externalID) {
+		// Can't finalize the transfer since it's already finalized
+		return false, nil
+	}
+
 	// Get the transfer info from the store to update it properly
-	transfer, err := k.GetInboundTransfer(ctx, externalID)
+	transfer, err := k.GetInboundTransfer(ctx, externalID, externalHeight)
 	switch {
 	case err == nil:
-	case errors.Is(err, ErrTransferNotFound):
+	case errors.Is(err, sdkerrors.ErrNotFound):
 		// If the transfer is new, then create it
-		transfer = types.NewInboundTransfer(externalID, destAddr, assetID, amount)
+		transfer = types.NewInboundTransfer(externalID, externalHeight, destAddr, assetID, amount)
 	default:
-		return fmt.Errorf("can't get the transfer info from store: %s", err.Error())
+		return false, fmt.Errorf("can't get the transfer info from store: %s", err.Error())
 	}
 
 	// Check if the sender has already signed this transfer
 	if slices.Contains(transfer.Voters, sender) {
-		return fmt.Errorf("the transfer has already been signed by this sender")
+		return false, fmt.Errorf("the transfer has already been signed by this sender")
 	}
```

Comment:
I think we are missing important logic here of ensuring that the sender is one of the validators from the superset

Reply:
Nope, it's done at the very beginning of the grpc method: the validateSenderIsSigner method. But I guess it would be more explicit to move it here

Reply:
Yeah I now notice that it was at the msg server level, but I like checking this at a higher level in case the method gets called outside of the msg server. Was probably fine as it was, I personally like this better for now